### PR TITLE
Add `HashEncoder` and `QueryEngineTests`

### PR DESCRIPTION
**Pull request by MaxDesiatov**
Previously, `QueryEngine` couldn't cache anything due to a missing persistable hashing implementation that could be consistent across SwiftPM relaunches. Here `HashEncoder` is implemented on top `Codable`. While it probably has certain performance overhead when compared to an implementation with macros, it should be good enough for prototyping.

New `QueryEngine` now works for trivial computations, as verified by newly added `QueryEngineTests`.

### CODE REVIEW

File: Tests/QueryEngineTests/QueryEngineTests.swift [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+  func write(_ path: FilePath, _ value: some Encodable) async throws {
+    let data = try encoder.encode(value)
+    try await self.withOpenWritableFile(path) { fileHandle in
+      try await fileHandle.write(data)
+    }
+  }
+}
+
+struct Const: Query {
```
Comment (MaxDesiatov) [inline Tests/QueryEngineTests/QueryEngineTests.swift:48]:
```suggestion
private struct Const: Query {
```

File: Tests/QueryEngineTests/QueryEngineTests.swift [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  let x: Int
+
+  func run(engine: QueryEngine) async throws -> FilePath {
+    let resultPath = FilePath("/Const-\(x)")
+    try await engine.fileSystem.write(resultPath, self.x)
+    return resultPath
+  }
+}
+
+struct MultiplyByTwo: Query {
```
Comment (MaxDesiatov) [inline Tests/QueryEngineTests/QueryEngineTests.swift:58]:
```suggestion
private struct MultiplyByTwo: Query {
```

File: Tests/QueryEngineTests/QueryEngineTests.swift [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    let constPath = try await engine[Const(x: self.x)].path
+    let constResult = try await engine.fileSystem.read(constPath, as: Int.self)
+
+    let resultPath = FilePath("/MultiplyByTwo-\(constResult)")
+    try await engine.fileSystem.write(resultPath, constResult * 2)
+    return resultPath
+  }
+}
+
+struct AddThirty: Query {
```
Comment (MaxDesiatov) [inline Tests/QueryEngineTests/QueryEngineTests.swift:71]:
```suggestion
private struct AddThirty: Query {
```

File: Tests/QueryEngineTests/QueryEngineTests.swift [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    let constPath = try await engine[Const(x: self.x)].path
+    let constResult = try await engine.fileSystem.read(constPath, as: Int.self)
+
+    let resultPath = FilePath("/AddThirty-\(constResult)")
+    try await engine.fileSystem.write(resultPath, constResult + 30)
+    return resultPath
+  }
+}
+
+struct Expression: Query {
```
Comment (MaxDesiatov) [inline Tests/QueryEngineTests/QueryEngineTests.swift:84]:
```suggestion
private struct Expression: Query {
```

File: Sources/QueryEngine/QueryEngine.swift [hunk 5]
```diff
@@ -34,13 +35,13 @@ public func withQueryEngine(
...
     private(set) var cacheMisses = 0
 
-    public let fileSystem: any AsyncFileSystem
-    public let httpClient = HTTPClient()
-    public let observabilityScope: ObservabilityScope
+    package let fileSystem: any AsyncFileSystem
+    package let httpClient = HTTPClient()
+    package let observabilityScope: ObservabilityScope
     private let resultsCache: SQLiteBackedCache<FileCacheRecord>
     private var isShutDown = false
```
Comment (MaxDesiatov) [inline Sources/QueryEngine/QueryEngine.swift:46]: Need to handle circular query dependencies here, probably with a dictionary from `Query` to query result continuation.

Comment (MaxDesiatov) [reply Sources/QueryEngine/QueryEngine.swift:46]: Don't recompute `Query` hash for deep FS trees.

File: Sources/QueryEngine/FileSystem/AsyncFileSystem.swift [hunk 6]
```diff
@@ -16,8 +16,15 @@ import struct SystemPackage.Errno
 import struct SystemPackage.FilePath
 
 public protocol AsyncFileSystem: Actor {
```
Comment (MaxDesiatov) [inline Sources/QueryEngine/FileSystem/AsyncFileSystem.swift:18]:
```suggestion
package protocol AsyncFileSystem: Actor {
```

File: Sources/QueryEngine/CacheKey.swift [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+@_exported import protocol Crypto.HashFunction
+import struct Foundation.URL
+import struct SystemPackage.FilePath
+
+/// Indicates that values of a conforming type can be hashed with an arbitrary hashing function. Unlike `Hashable`,
+/// this protocol doesn't utilize random seed values and produces consistent hash values across process launches.
+package protocol CacheKey: Encodable {
```
Comment (MaxDesiatov) [inline Sources/QueryEngine/CacheKey.swift:19]: Comment about macros not usable in this case because of Swift 5.8 and CMake requirements.

**Review by xedin — APPROVED** [review]

**Comment by MaxDesiatov** [comment]
@swift-ci test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows
@swift-ci test windows

**Pull request merged**